### PR TITLE
lib: more cache fixes

### LINKED ISSUE #3806: The cache handler should not invalidate all requests to an origin
When the cache-handler recevies a succesful response from one of "unsafe" http methods such as POST/PUT/DELETE... it invalidates _all_ other cached responses to this origin. 

https://github.com/nodejs/undici/blob/main/lib/handler/cache-handler.js#L79

RFC says:

`
A cache MUST invalidate the target URI ([Section 7.1](https://www.rfc-editor.org/rfc/rfc9110#section-7.1) of [[HTTP](https://www.rfc-editor.org/rfc/rfc9111.html#HTTP)]) when it receives a non-error status code in response to an unsafe request method (including methods whose safety is unknown).
`

By "target URI" here I understand the the full URL and not just an origin (correct me here if I;m wrong). For example if a client receives a succesful response for the `DELETE http://example.com/foo/bar` the cache hanlder can delete the `GET http://example.com/foo/bar`, but not all other cached responses to the `http://example.com`.

IMHO With a current behaviour the cache seems useless, because a one regulary called POST requests keeps your cache almost always empty.

### PR DESCRIPTION
**Pull request by flakey5**
Closes #3806

- Removes `CacheStoreReadable` and `CacheStoreWritable`, replaces w/ just `Readable` and `Writable`
  - Reasoning: api simplicity
- Renames `CacheStoreValue` type to `CachedResponse`
- Removes `CacheStore#createReadStream` function (see one below on reasoning)
- Adds `CacheStore#get` function that acts similarly to `createReadStream`
  - Reasoning: api simplicity & niceness. Returns a `GetResult` which separates the `CacheResponse` and `Readable` for the body
- Simplifies `MemoryCacheStore`
  - Inline readable & writables
  - `readLock` and `writeLock` condensed into just `locked` that only stops value from being read
- Introduces `CacheKey` type that we pass into the cache handler, cache stores
  - Replaces us passing in the full request options
  - Reasoning: cache stores don't need the full thing, guides them to be more spec-compliant
- Various internal type fixes
- Fixes regarding a header being present multiple times
  - Stripping headers would result in them being re-encoded in the wrong format
  - Cache control & vary header parsings
- Makes `isFull` optional
  - Reasoning: some cache stores can't get full / don't have a max. This only exists as a very small optimization
- Removes caching of trailers
  - Reasoning: simplicity sakes, ideally this will be added back later
- Replace `deleteByOrigin` with `deleteByKey`
  - Reasoning: #3806
- Adds `CachedValue` type
  - Separates the `body` stream and the properties in `CachedResponse`
  - Allows the stream to be undefined for HEAD requests since there's no body

cc @mcollina @ronag

### CODE REVIEW

File: docs/docs/api/CacheStore.md [hunk 1]
```diff
@@ -21,27 +22,32 @@ The store must implement the following functions:
 
 ### Getter: `isFull`
 
-This tells the cache interceptor if the store is full or not. If this is true,
+Optional. This tells the cache interceptor if the store is full or not. If this is true,
 the cache interceptor will not attempt to cache the response.
 
-### Function: `createReadStream`
+### Function: `getValueByKey`
```
Comment (ronag) [inline docs/docs/api/CacheStore.md:28]: Any reason it can't be just `get`?

File: lib/handler/cache-handler.js [hunk 2]
```diff
@@ -176,7 +175,7 @@ class CacheHandler extends DecoratorHandler {
   onData (chunk) {
     let paused = false
 
-    if (this.#writeStream) {
+    if (this.#writeStream && this.#cacheKey.method !== 'HEAD') {
```
Comment (ronag) [inline lib/handler/cache-handler.js:178]: If HEAD then shouldn't writeStream just be null?

File: lib/handler/cache-handler.js [hunk 3]
```diff
@@ -295,7 +297,7 @@ function determineStaleAt (now, headers, cacheControlDirectives) {
     return now + (maxAge * 1000)
   }
 
-  if (headers.expire) {
+  if (headers.expire && typeof headers.expire === 'string') {
     // https://www.rfc-editor.org/rfc/rfc9111.html#section-5.3
     return now + (Date.now() - new Date(headers.expire).getTime())
```
Comment (ronag) [inline lib/handler/cache-handler.js:302]: Might want to make sure expire is a valid date?

File: lib/interceptor/cache.js [hunk 4]
```diff
@@ -47,27 +47,30 @@ module.exports = (opts = {}) => {
...
+       * @param {import('../../types/cache-interceptor.d.ts').default.CachedResponse} value
        */
       const respondWithCachedValue = (stream, value) => {
-        assert(!stream.destroyed, 'stream should not be destroyed')
-        assert(!stream.readableDidRead, 'stream should not be readableDidRead')
-
+        assert(!stream || !stream.destroyed, 'stream should not be destroyed')
+        assert(!stream || !stream.readableDidRead, 'stream should not be readableDidRead')
         try {
           stream
```
Comment (ronag) [inline lib/interceptor/cache.js:72]: Can stream be nully? When?

File: lib/interceptor/cache.js [hunk 5]
```diff
@@ -76,21 +79,22 @@ module.exports = (opts = {}) => {
...
+                handler.onComplete([])
               }
             })
 
           if (typeof handler.onConnect === 'function') {
             handler.onConnect((err) => {
-              stream.destroy(err)
+              if (stream) {
+                stream.destroy(err)
+              }
```
Comment (ronag) [inline lib/interceptor/cache.js:94]:
```suggestion
              stream?.destroy(err)
```

File: lib/interceptor/cache.js [hunk 6]
```diff
@@ -103,13 +107,15 @@ module.exports = (opts = {}) => {
...
+            if (handler.onHeaders(value.statusCode, rawHeaders, () => stream?.resume(), value.statusMessage) === false) {
+              stream?.pause()
             }
           }
 
           if (opts.method === 'HEAD') {
-            stream.destroy()
+            if (typeof handler.onComplete === 'function') {
+              handler.onComplete([])
+            }
```
Comment (ronag) [inline lib/interceptor/cache.js:118]: either destroy stream or `assert(!stream)`

File: lib/interceptor/cache.js [hunk 7]
```diff
@@ -162,22 +168,23 @@ module.exports = (opts = {}) => {
...
+      if (typeof result.then === 'function') {
+        result.then((result) => {
+          if (!result) {
+            // Request isn't cached
+            return dispatch(opts, new CacheHandler(globalOpts, cacheKey, handler))
           }
-        })
+
+          handleStream(result)
+        }).catch(handler.onError)
```
Comment (ronag) [inline lib/interceptor/cache.js:185]:
```suggestion
        }).catch(err => handler.onError(err))
```

Comment (ronag) [reply lib/interceptor/cache.js:185]: If handleStream throws we should crash, that should never throw...

File: lib/util/cache.js [hunk 8]
```diff
@@ -4,6 +4,31 @@ const {
...
+   */
+  const cacheKey = {
+    origin: opts.origin.toString(),
+    method: opts.method,
+    path: opts.path
+  }
+
+  if (opts.headers) {
+    cacheKey.headers = opts.headers
+  }
```
Comment (ronag) [inline lib/util/cache.js:27]:
```suggestion
  const cacheKey = {
    origin: opts.origin.toString(),
    method: opts.method,
    path: opts.path,
    headers: opts.headers
  }
```

File: lib/core/util.js [hunk 9]
```diff
@@ -516,6 +478,24 @@ function parseRawHeaders (headers) {
...
+    throw new TypeError('expected headers to be an array')
+  }
+
+  const rawHeaders = new Array(headers.length)
+
+  for (let i = 0; i < headers.length; i++) {
+    rawHeaders[i] = Buffer.from(headers[i])
+  }
+
+  return rawHeaders
```
Comment (ronag) [inline lib/core/util.js:496]:
```suggestion
  return headers.map(x => Buffer.from(x))
```

File: lib/interceptor/cache.js (hunk 4 above)
Comment (flakey5) [reply lib/interceptor/cache.js:72]: On HEAD requests

File: docs/docs/api/CacheStore.md (hunk 1 above)
Comment (flakey5) [reply docs/docs/api/CacheStore.md:28]: Personal preference, but either or work for me

File: lib/interceptor/cache.js (hunk 4 above)
Comment (KhafraDev) [reply lib/interceptor/cache.js:72]: the asserts already catch it

File: docs/docs/api/CacheStore.md (hunk 1 above)
Comment (ronag) [reply docs/docs/api/CacheStore.md:28]: Seem unnecessarily verbose to me...

File: lib/interceptor/cache.js (hunk 4 above)
Comment (ronag) [reply lib/interceptor/cache.js:72]: This is broken though like this... onConnect cannot abort the request

**Comment by DTrombett** [comment]
Asking here as it seems related, but why is the `public` directive required? Shouldn't it be required only with authenticated requests?
https://github.com/nodejs/undici/blob/c94fda14b0c58cd1083c6efb7a49868bab4d835f/lib/handler/cache-handler.js#L238-L245

**Comment by flakey5** [comment]
> Asking here as it seems related, but why is the public directive required? Shouldn't it be required only with authenticated requests?

Will fix, thanks!

File: lib/interceptor/cache.js (hunk 4 above)
Comment (flakey5) [reply lib/interceptor/cache.js:72]: Wdym?

**Review by ronag — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**